File: 0347-TopKFrequentElements/Solution.java
import java.util.*;

public class Solution {
    
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> numToCountMap = new HashMap<>();
        for (int num : nums) {
            numToCountMap.put(num, numToCountMap.getOrDefault(num, 0) + 1);
        }
        
        Map<Integer, Set<Integer>> countToNumsMap = new HashMap<>();
        int maxCount = 0;
        for (Map.Entry<Integer, Integer> e : numToCountMap.entrySet()) {
            int num = e.getKey();
            int count = e.getValue();
            Set<Integer> set = countToNumsMap.get(count);
            if (set == null) {
                set = new HashSet<>();
                countToNumsMap.put(count, set);
            }
            set.add(num);
            maxCount = Math.max(maxCount, count);
        }

        int[] result = new int[k];
        int i = 0;
        int j = maxCount;
        while (i < k && j >= 0) {
            if (countToNumsMap.containsKey(j)) {
                for (int num : countToNumsMap.get(j)) {
                    result[i++] = num;
                }
            }
            j--;
        }
        return result;
    }

    public static void main(String[] args) {
        System.out.println();
    }

}

File: 0347-TopKFrequentElements/solution.py
class Solution:

    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        c = Counter(nums)

        bins = [None for _ in range(len(nums))]
        for key, val in c.items():
            if not bins[val - 1]:
                bins[val - 1] = []
            bins[val - 1].append(key)

        res = []
        for i in range(len(nums) - 1, -1, -1):
            curr_bin = bins[i]
            if curr_bin:
                res += curr_bin
                k -= len(curr_bin)
                if k == 0:
                    break
        return res


if __name__ == "__main__":
    s = Solution()

